scio-parquet/src/main/scala/com/spotify/scio/parquet/GcsConnectorUtil.scala

```diff
@@ -17,33 +17,52 @@
 
 package com.spotify.scio.parquet
 
+import com.google.auth.oauth2.{GoogleCredentials, ServiceAccountCredentials}
+import com.google.cloud.hadoop.util.AccessTokenProvider
 import com.spotify.scio.ScioContext
 import com.spotify.scio.util.ScioUtil
+import org.apache.hadoop.conf.Configuration
 import org.apache.hadoop.mapreduce.Job
 import org.apache.hadoop.mapreduce.lib.input.FileInputFormat
 
+import java.io.File
+import java.util.Locale
+import scala.util.Try
+
 private[parquet] object GcsConnectorUtil {
-  def setCredentials(job: Job): Unit =
+  def setCredentials(job: Job): Unit = {
     // These are needed since `FileInputFormat.setInputPaths` validates paths locally and
     // requires the user's GCP credentials.
-    sys.env.get("GOOGLE_APPLICATION_CREDENTIALS") match {
-      case Some(json) =>
-        job.getConfiguration
-          .set("fs.gs.auth.service.account.json.keyfile", json)
-      case None =>
-        // Client id/secret of Google-managed project associated with the Cloud SDK
+    lazy val applicationDefault = Try(GoogleCredentials.getApplicationDefault())
+
+    sys.env
+      .get("GOOGLE_APPLICATION_CREDENTIALS")
+      .filter(_.nonEmpty)
+      .orElse {
+        applicationDefault.toOption match {
+          case Some(_: ServiceAccountCredentials) => getWellKnownCredentialFile.map(_.toString)
+          case _                                  => None
+        }
+      } match {
+      case Some(serviceAccountKeyFile) =>
+        job.getConfiguration.set("fs.gs.auth.service.account.json.keyfile", serviceAccountKeyFile)
+      case _ if applicationDefault.isSuccess => // ADC exists but isn't a service account
```

Comment:
I would prefer this to be more explicit / easier to read with smth like:
```scala
      case None =>
        Try(GoogleCredentials.getApplicationDefault()) match {
          case Success(_: ServiceAccountCredentials) =>
            job.getConfiguration.set(
              "fs.gs.auth.access.token.provider.impl",
              "com.spotify.scio.parquet.ApplicationDefaultTokenProvider"
            )
          case _ =>
            job.getConfiguration
              .setBoolean("fs.gs.auth.service.account.enable", false)
            job.getConfiguration.unset("fs.gs.auth.null.enable")
        }
```
this avoids the `.orElse` above

Reply:
that is cleaner to read! It changes the logic a little bit though... in my version if either `GOOGLE_APPLICATION_CREDENTIALS` or `~/.config/gcloud/application_default_credentials.json` exists (_and_ it's a service account credential), then we set `fs.gs.auth.service.account.json.keyfile`. Else, if `~/.config/gcloud/application_default_credentials.json` exists (and it's not a service account credential) then we set `fs.gs.auth.access.token.provider.impl`. So, the reason I had factored out `Try(GoogleCredentials.getApplicationDefault())` into a local variable was to avoid evaluating it twice.

Reply:
Ahah I see yes I did miss that! 😄 my comment edit skills are not yet there 😄

Reply:
So keeping the prev line of thought we just need to add a `map` there. So smth like this (didn't test).

```
        Try(GoogleCredentials.getApplicationDefault()).map {
          case _: ServiceAccountCredentials => getWellKnownCredentialFile.map(_.toString)
          case _                            => None
        } match {
          case Success(Some(sa)) =>
            job.getConfiguration.set("fs.gs.auth.service.account.json.keyfile", sa) // it's ok to be duped because here it exposes the intent.
          case Success(None) =>
            job.getConfiguration.set(
              "fs.gs.auth.access.token.provider.impl",
              "com.spotify.scio.parquet.ApplicationDefaultTokenProvider"
            )
          case _ =>
            job.getConfiguration.setBoolean("fs.gs.auth.service.account.enable", false)
            job.getConfiguration.unset("fs.gs.auth.null.enable")
        }
```

Reply:
cool, I like that! it does remove the `sys.env.get("GOOGLE_APPLICATION_CREDENTIALS")` call on our end but, that's fine because that's the first thing `GoogleCredentials.getApplicationDefault()` does anyway. However in that case I should also update the `getWellKnownCredentialFile` method to check for a file existing at `$GOOGLE_APPLICATION_CREDENTIALS` location first 👍